### PR TITLE
add function to validate quantities in ntuple, while loading

### PR DESCRIPTION
**Pull request by harrypuuter**
This will highlight files that do not contain the same quantities as the first file checked. In rare cases, where the first file is wrong, every other file will appear as a buggy one.

### CODE REVIEW

File: booking.py [hunk 1]
```diff
@@ -206,6 +207,35 @@ def add_tagged_friends(friends):
...
+        friend_quantitites = set()
+        for f in friends:
+            friend = TFile(f)
+            friend_quantitites.update(set([x.GetName() for x in friend.Get("ntuple").GetListOfLeaves()]))
+        # first we check the main ntuple, then the friends
+        errordata = {}
+        if len(validation_dict["varset"]) == 0:
+            validation_dict["varset"] = quantities
+        else:
+            difference = (validation_dict["varset"] - quantities) | (quantities - validation_dict["varset"])
```
Comment (mburkart) [inline booking.py]: Why not use symmetric difference here?
```suggestion
            difference = validation_dict["varset"] ^ quantities
```

File: booking.py [hunk 2]
```diff
@@ -206,6 +207,35 @@ def add_tagged_friends(friends):
...
+            if len(difference) != 0:
+                # error is found
+                errordata["file"] = root_file_path
+                errordata["difference"] = difference
+        if len(validation_dict["friends_varset"]) == 0:
+            validation_dict["friends_varset"] = friend_quantitites
+        else:
+            difference = (validation_dict["friends_varset"] - friend_quantitites) | (friend_quantitites - validation_dict["friends_varset"])
```
Comment (mburkart) [inline booking.py:230]: See above

File: booking.py [hunk 3]
```diff
@@ -233,6 +273,22 @@ def add_tagged_friends(friends):
                 friends.append(Ntuple(friend_path, tdf_tree_friend))
         if not is_empty_file(root_file, tdf_tree):
             ntuples.append(Ntuple(root_file, tdf_tree, add_tagged_friends(friends)))
+            if validate_samples:
+                check_validity(root_file, validation_dict, friend_paths)
+    if len(validation_dict["errors"]) != 0:
+        logger.fatal("Validation for {} - {} - {} failed, differences were found".format(era, channel, dataset_name))
```
Comment (mburkart) [inline booking.py:279]: Does the code below this line run? I am not sure if `logger.fatal` raises an error anymore.

File: booking.py (hunk 2 above)
Comment (mburkart) [inline booking.py]: See above
```suggestion
            difference = validation_dict["friends_varset"] ^ friend_quantitites
```

File: booking.py (hunk 1 above)
Comment (harrypuuter) [reply booking.py]: i did not know about this one ! Changed

File: booking.py (hunk 3 above)
Comment (harrypuuter) [reply booking.py:279]: Logger Fatal just results in something like this:
```
ntuple_processor.booking - CRITICAL - Validation for 2018 - mt - qqH failed, differences were found
ntuple_processor.booking - CRITICAL - File /ceph/sbrommer/smhtt_ul/2022_07_v6/ntuples/2018/ZHToTauTau_M125_CP5_13TeV-powheg-pythia8_RunIISummer20UL18NanoAODv9-106X/mt/ZHToTauTau_M125_CP5_13T
eV-powheg-pythia8_RunIISummer20UL18NanoAODv9-106X_5.root has the following differences:
ntuple_processor.booking - CRITICAL -   {'THU_qqH_Mjj1000', 'THU_qqH_Mjj120', 'THU_qqH_TOT', 'THU_qqH_Mjj60', 'THU_qqH_Mjj350', 'THU_qqH_PTH200', 'THU_qqH_Mjj1500', 'THU_qqH_JET01', 'THU_qqH
_Mjj700', 'THU_qqH_25'}
```
but still continues running

**Review by mburkart — APPROVED** [review]
Everything looks good!

**Pull request merged**